src/harness/test_harness_objects.py

```diff
@@ -0,0 +1,299 @@
+      by the caller(
+      heartbeat_responses: list of heart beat response
+    """
+    # Concatenate all the heartbeat request of CBSD into single request
+    heartbeat_requests = []
+    for cbsd_object_item in self.cbsd_objects.values():
+      heartbeat_requests = cbsd_object_item.constructHeartbeatRequestForAllActiveGrants()
+
+    if len(heartbeat_requests):
+      heartbeat_requests_wrapped = {
+        'heartbeatRequest': heartbeat_requests
+      }
+      # Perform heartbeat requests
+      heartbeat_responses = self.testcase._sas.Heartbeat(
+        heartbeat_requests_wrapped, self.ssl_cert, self.ssl_key)['heartbeatResponse']
+
+      for heartbeat_response in heartbeat_responses:
+        try:
+          cbsd_object = self.cbsd_objects[heartbeat_response['cbsdId']]
+          grant_object = cbsd_object.grant_objects[heartbeat_response['grantId']]
+        except KeyError:
+          logging.debug("CBSD or the Grant object not found")
+
+        if heartbeat_response['response']['responseCode'] == 0:
+          grant_object.authorized_in_last_heartbeat = True
+        else:
+          self.mapResponseCodeToGrantState(heartbeat_response['response']['responseCode'],
+                                           grant_object)
+      return heartbeat_requests, heartbeat_responses
+
+  def mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+    """Maps the heartbeat response and update the grant state"""
+    if heartbeat_response_code == 502:
+      grant_object.is_terminated = True
+    elif heartbeat_response_code == 500:
+      grant_object.is_terminated = True
+      grant_object.authorized_in_last_heartbeat = False
+    elif heartbeat_response_code == 501:
+      grant_object.authorized_in_last_heartbeat = False
+
+  def performHeartbeatAndUpdateGrants(self):
+    """Does heartbeat request for all active grants and updates the grants.
+
+    The following steps are performed by this function
+      1. Send heartbeat request on all grant of CBSD at once
+      2. If the heartbeat response contains operationParam then those grants
+         are relinquished.
+      3. New grant requests with the suggested operationParam values in the
+         heartbeat response is sent. Failures are ignored.
+      4. Heartbeat is performed again for all grants. Failures are ignored.
+    """
+    # Heartbeat on all active grants
+    heartbeat_requests = []
+    heartbeat_requests, heartbeat_responses = self.heartbeatRequestForAllGrants()
+    relinquish_requests = []
+    grant_requests = []
+
+    # Construct relinquishment and grant for the heart beat response which have operation param
+    for heartbeat_response, heartbeat_request in zip(heartbeat_responses, heartbeat_requests):
+        if heartbeat_response['response']['responseCode'] != 0:
+          # If any heartbeats contain suggested operation params,construct relinquish requests
+          if 'operationParam' in heartbeat_response:
```

Comment:
if response code is TERMINATED_GRANT and operationParam is not present then raise or log an error. As the response is invalid.

Reply:
Added   error log as per suggestion